File: Golang/387. First Unique Character in a String.go
// func firstUniqChar(s string) int { // 44ms
//     countChars := make(map[int]int, 26)
//     for _, char := range s {
//         countChars[int(char) - 'a']++
//     }
//     for i, char := range s {
//         if countChars[int(char) - 'a'] == 1 {
//             return i
//         }
//     }
//     return -1
// }

// func firstUniqChar(s string) int { // 24ms 既然用了ascii码就该用数组
//     countChars := [26]int{}
//     for _, char := range s {
//         countChars[int(char) - 'a']++
//     }
//     for i, char := range s {
//         if countChars[int(char) - 'a'] == 1 {
//             return i
//         }
//     }
//     return -1
// }

func firstUniqChar(s string) int { // 12ms 超快，但是占用更多空间，虽然依然是O(1)的space
    countChars := [256]int{}
    for _, char := range s {
        countChars[int(char)]++
    }
    for i, char := range s {
        if countChars[int(char)] == 1 {
            return i
        }
    }
    return -1
}

File: Python3/387. First Unique Character in a String.py
class Solution:
    def firstUniqChar(self, s):
        """
        :type s: str
        :rtype: int
        """
        # countChars = {} # 160ms
        # for char in s:
        #     countChars[char] = countChars.get(char, 0) + 1
        # for index, char in enumerate(s):
        #     if countChars[char] == 1:
        #         return index
        # return -1
        
        # countChars = collections.Counter(s) # 124ms
        # for index, char in enumerate(s):
        #     if countChars[char] == 1:
        #         return index
        # return -1
        
        letters = [chr(i) for i in range(ord('a'), ord('z')+1)] # 88ms
        index = [s.index(l) for l in letters if s.count(l) == 1]
        return min(index) if len(index) > 0 else -1
